### ISSUE TITLE
Questions about the `transform_eval` in panoptic_inference (panoptic segmentation post-processing)

### ISSUE THREAD

**Issue by Li-Qingyun** [IDEA-Research/MaskDINO#59]
Hi, thanks for your great work! 
I'm trying to understand the method and implement it with mmdet 3.0 in open-mmlab/mmdetection#9808. (as a student contributor, not openmmlab stuff)

I have questions about the `transform_eval` in [panoptic_inference](https://github.com/IDEA-Research/MaskDINO/blob/52283032ed76fd7f8a399a1f337cc617cf3fb44f/maskdino/maskdino.py#L393-L453), althougn there has been a comment here tring to tell the users about the option.

In Mask DINO:
```Python3
# As we use focal loss in training, evaluate with sigmoid. As sigmoid is mainly for detection and not sharp
# enough for semantic and panoptic segmentation, we additionally use use softmax with a temperature to
# make the score sharper.
scores, labels = mask_cls.sigmoid().max(-1)
mask_pred = mask_pred.sigmoid()
keep = labels.ne(self.sem_seg_head.num_classes) & (scores > self.object_mask_threshold)  # 0.25 now
# added process
if self.transform_eval:
    scores, labels = F.softmax(mask_cls.sigmoid() / T, dim=-1).max(-1)
cur_scores = scores[keep]
cur_classes = labels[keep]
cur_masks = mask_pred[keep]
cur_prob_masks = cur_scores.view(-1, 1, 1) * cur_masks
```

In [Mask2Former](https://github.com/facebookresearch/Mask2Former)
The [correspondng code](https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/mask2former/maskformer_model.py#L286-L342) is:
```Python3
scores, labels = F.softmax(mask_cls, dim=-1).max(-1)
mask_pred = mask_pred.sigmoid()
keep = labels.ne(self.sem_seg_head.num_classes) & (scores > self.object_mask_threshold)  # 0.8
cur_scores = scores[keep]
cur_classes = labels[keep]
cur_masks = mask_pred[keep]
cur_mask_cls = mask_cls[keep]
cur_mask_cls = cur_mask_cls[:, :-1]
cur_prob_masks = cur_scores.view(-1, 1, 1) * cur_masks
```

I have understood that, the Focal Loss is used, hence sigmoid should be used to obtain the scores.
1. what's the meaning of `sharp`? what is the score like if it is `sharp`? Can you give descriptions and provide some examples?
2. why the high-scoring `keep` is still obtained with 'not sharp' sigmoid. why the softmax with temperature is not used globally?
3. for the softmax with temperature, can you share more designing reasons about the formula `softmax(mask_cls.sigmoid() / T)`? both for panoptic and for semantic.
4. i have not understood how the following logic about `overlap_threshold` works, although there has been [closed issue](https://github.com/facebookresearch/Mask2Former/issues/132) about it.
> Used to filter a prediction that is occluded by a prediction with a higher confidence score by a large area

but does the `mask_area / original_area` have certain specific meaning that can be understood.
can you share your understandings?
```Python3
cur_prob_masks = cur_scores.view(-1, 1, 1) * cur_masks
......
cur_mask_ids = cur_prob_masks.argmax(0)

for k in range(cur_classes.shape[0]):
    pred_class = cur_classes[k].item()
    isthing = pred_class in self.metadata.thing_dataset_id_to_contiguous_id.values()
    mask_area = (cur_mask_ids == k).sum().item()
    original_area = (cur_masks[k] >= prob).sum().item()
    mask = (cur_mask_ids == k) & (cur_masks[k] >= prob)

    if mask_area > 0 and original_area > 0 and mask.sum().item() > 0:
        if mask_area / original_area < self.overlap_threshold:
            continue
        ......
```
My questions seem long, thanks for the authors for sharing.

Thanks Thanks and Thanks!

**Comment by Li-Qingyun** [comment]
@HaoZhang534 @FengLi-ust

**Comment by FengLi-ust** [comment]
Because the score distribution of softmax and sigmoid are different, i.e, softmax is more like one-hot, where only one logit is high and close to 1, other logits are close to 0; while sigmoid just normalizes each logit into [0,1] without considering other logits.

**Comment by Li-Qingyun** [comment]
@FengLi-ust OK, thanks for your quick reply!
